File: ядро/образ_в_атомы.py
# -*- coding: utf-8
"""Обратная проекция: облако точек образа → скелеты атомов.

Инверсия проекция_образ.js (режим «физика»):
  X = время, Y = log10(freq), Z = amp.

Канон (ТЗ §4.2):
  birth = norm(x) · T_образа
  freq  = exp(norm(y) · [ln 60 .. ln 9000])
  amp   = norm(z) → [0.05 .. 1.0]
"""
from __future__ import annotations

import json
import math
import os
from typing import Any, Mapping, Sequence

import numpy as np

from ядро.пороги import BAND_F_HI, BAND_F_LO

КОРЕНЬ = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
КЛЕТКИ = os.path.join(КОРЕНЬ, "данные", "клеточки")

# Масштабы сцены — зеркало проекция_образ.js (режим физика)
X_SCALE = 1.8
Y_SCALE = 2.4
Z_SCALE = 1.2
T_ОБРАЗА_ПО_УМОЛЧ = 3.0
AMP_MIN, AMP_MAX = 0.05, 1.0
ОКНО_BIRTH = 0.05
MAX_СКЕЛЕТОВ_В_ОКНЕ = 24
K_СОСЕДЕЙ = 8


def _clip01(v: float) -> float:
    return float(max(0.0, min(1.0, v)))


def _norm(v: float, a: float, b: float) -> float:
    if abs(b - a) < 1e-12:
        return 0.5
    return (v - a) / (b - a)


def _границы_атомов(atoms: Sequence[Mapping]) -> dict[str, float]:
    births = [float(a.get("birth") or 0) for a in atoms]
    lfs = [math.log10(max(1.0, float(a.get("freq") or 20))) for a in atoms]
    amps = [float(a.get("amp") or 0) for a in atoms]
    return {
        "tmin": min(births),
        "tmax": max(births),
        "fmin": min(lfs),
        "fmax": max(lfs),
        "amax": max(max(amps), 1e-6),
    }


def атомы_в_облако(
    atoms: Sequence[Mapping],
    *,
    t_образа: float | None = None,
    канон: bool = False,
) -> tuple[list[dict], dict]:
    """Прямая проекция атомов → облако (для тестов round-trip)."""
    if not atoms:
        return [], {}
    b = _границы_атомов(atoms)
    if канон:
        meta = {"режим": "канон", "t_образа": t_образа or T_ОБРАЗА_ПО_УМОЛЧ}
    else:
        meta = {**b, "режим": "физика", "t_образа": t_образа or T_ОБРАЗА_ПО_УМОЛЧ}
    облако = []
    for a in atoms:
        birth = float(a.get("birth") or 0)
        freq = max(1.0, float(a.get("freq") or 20))
        amp = float(a.get("amp") or 0)
        if канон:
            nx = _clip01(birth / meta["t_образа"])
            x = (2 * nx - 1) * X_SCALE
            ny = _clip01((math.log(freq) - math.log(BAND_F_LO)) / (math.log(BAND_F_HI) - math.log(BAND_F_LO)))
            y = (ny - 0.5) * Y_SCALE
            nz = _clip01((amp - AMP_MIN) / (AMP_MAX - AMP_MIN))
            z = nz * Z_SCALE
        else:
            x = (2 * _norm(birth, b["tmin"], b["tmax"]) - 1) * X_SCALE
            y = (_norm(math.log10(freq), b["fmin"], b["fmax"]) - 0.5) * Y_SCALE
            z = (amp / b["amax"]) * Z_SCALE
        pt = {
            "x": round(x, 6),
            "y": round(y, 6),
            "z": round(z, 6),
            "size": float(a.get("size") or max(0.05, amp)),
        }
        if a.get("color_r") is not None:
            pt["r"] = int(a["color_r"])
            pt["g"] = int(a.get("color_g") or 0)
            pt["b"] = int(a.get("color_b") or 0)
        облако.append(pt)
    return облако, meta


def _точка_в_скелет(
    pt: Mapping,
    meta: Mapping,
) -> dict[str, float]:
    x, y, z = float(pt["x"]), float(pt["y"]), float(pt["z"])
    w = float(pt.get("size") or 1.0)

    if meta.get("режим") == "физика" and "tmin" in meta:
        birth = (x / X_SCALE + 1) / 2
        birth = birth * (float(meta["tmax"]) - float(meta["tmin"])) + float(meta["tmin"])
        logf = y / Y_SCALE + 0.5
        logf = logf * (float(meta["fmax"]) - float(meta["fmin"])) + float(meta["fmin"])
        freq = 10.0 ** logf
        amp = (z / Z_SCALE) * float(meta["amax"])
        amp = max(0.0, amp)
    else:
        t = float(meta.get("t_образа") or T_ОБРАЗА_ПО_УМОЛЧ)
        nx = _clip01((x / X_SCALE + 1) / 2)
        birth = nx * t
        ny = _clip01(y / Y_SCALE + 0.5)
        ln_lo, ln_hi = math.log(BAND_F_LO), math.log(BAND_F_HI)
        freq = math.exp(ny * (ln_hi - ln_lo) + ln_lo)
        nz = _clip01(z / Z_SCALE)
        amp = AMP_MIN + nz * (AMP_MAX - AMP_MIN)

    amp_out = amp if meta.get("режим") == "физика" else max(AMP_MIN, min(AMP_MAX, amp))
    sk = {
        "birth": round(max(0.0, birth), 6),
        "freq": round(max(BAND_F_LO, min(BAND_F_HI, freq)), 4),
        "amp": round(amp_out, 6),
        "_w": w,
    }
    if "r" in pt:
        sk["цвет"] = {"r": pt["r"], "g": pt.get("g"), "b": pt.get("b")}
    return sk


def _слить_окно(group: list[dict]) -> dict:
    wsum = sum(g["_w"] for g in group) or 1.0
    out = {
        "birth": sum(g["birth"] * g["_w"] for g in group) / wsum,
        "freq": sum(g["freq"] * g["_w"] for g in group) / wsum,
        "amp": sum(g["amp"] * g["_w"] for g in group) / wsum,
    }
    if group[0].get("цвет"):
        out["цвет"] = group[0]["цвет"]
    return out


def _ограничить_плотность(скелеты: list[dict]) -> list[dict]:
    """≤ MAX_СКЕЛЕТОВ_В_ОКНЕ на каждые ОКНО_BIRTH секунд."""
    if not скелеты:
        return []
    sorted_s = sorted(скелеты, key=lambda s: s["birth"])
    out: list[dict] = []
    i = 0
    n = len(sorted_s)
    while i < n:
        t0 = sorted_s[i]["birth"]
        group = []
        while i < n and sorted_s[i]["birth"] - t0 <= ОКНО_BIRTH + 1e-9:
            group.append(sorted_s[i])
            i += 1
        if len(group) <= MAX_СКЕЛЕТОВ_В_ОКНЕ:
            out.extend(group)
        else:
            group.sort(key=lambda g: g["_w"], reverse=True)
            keep = group[:MAX_СКЕЛЕТОВ_В_ОКНЕ]
            merged_rest = [_слить_окно(group[MAX_СКЕЛЕТОВ_В_ОКНЕ:])] if len(group) > MAX_СКЕЛЕТОВ_В_ОКНЕ else []
            out.extend(keep + merged_rest)
    return out


def _лок_кривизна(pts: np.ndarray, i: int, k: int = K_СОСЕДЕЙ) -> float:
    n = len(pts)
    if n < 3:
        return 0.0
    d = np.linalg.norm(pts - pts[i], axis=1)
    d[i] = np.inf
    nn = np.argsort(d)[: min(k, n - 1)]
    block = pts[nn]
    if len(block) < 2:
        return 0.0
    c = block - block.mean(axis=0)
    cov = c.T @ c / max(len(block) - 1, 1)
    evals = np.linalg.eigvalsh(cov)
    evals = np.sort(np.maximum(evals, 0))[::-1]
    s = float(evals.sum())
    return float(evals[-1] / s) if s > 1e-12 else 0.0


def _лок_плотность(pts: np.ndarray, i: int, k: int = K_СОСЕДЕЙ) -> float:
    n = len(pts)
    if n <= 1:
        return 1.0
    d = np.linalg.norm(pts - pts[i], axis=1)
    d[i] = np.inf
    nn = np.sort(d)[: min(k, n - 1)]
    r = float(np.mean(nn)) if len(nn) else 1.0
    return round(1.0 / max(r, 1e-3), 4)


def облако_в_скелеты(
    облако: Sequence[Mapping],
    *,
    meta: Mapping | None = None,
    t_образа: float = T_ОБРАЗА_ПО_УМОЛЧ,
    ограничить_плотность: bool = True,
) -> list[dict]:
    """Облако → скелеты {birth, freq, amp, лок_кривизна, лок_плотность, цвет?}."""
    if not облако:
        return []
    m = dict(meta or {})
    if "режим" not in m:
        m.update({"режим": "канон", "t_образа": t_образа})

    raw = [_точка_в_скелет(p, m) for p in облако]
    if ограничить_плотность:
        raw = _ограничить_плотность(raw)

    pts = np.array([[s["birth"], math.log(max(s["freq"], 1)), s["amp"]] for s in raw], dtype=np.float64)
    out = []
    for i, s in enumerate(raw):
        sk = {
            "birth": s["birth"],
            "freq": s["freq"],
            "amp": s["amp"],
            "лок_кривизна": round(_лок_кривизна(pts, i), 6),
            "лок_плотность": _лок_плотность(pts, i),
        }
        if s.get("цвет"):
            sk["цвет"] = s["цвет"]
        out.append(sk)
    return out


def образ_в_атомы(
    облако: Sequence[Mapping],
    *,
    meta: Mapping | None = None,
    t_образа: float = T_ОБРАЗА_ПО_УМОЛЧ,
) -> list[dict]:
    """Публичный API — то же, что облако_в_скелеты."""
    return облако_в_скелеты(облако, meta=meta, t_образа=t_образа)


def из_json_облака(path: str) -> list[dict]:
    d = json.load(open(path, encoding="utf-8"))
    if isinstance(d, list):
        return list(d)
    return list(d.get("точки") or d.get("points") or [])


def из_obj(path: str) -> list[dict]:
    """Минимальный OBJ → облако {x,y,z}."""
    verts = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            if line.startswith("v "):
                parts = line.split()
                if len(parts) >= 4:
                    verts.append({"x": float(parts[1]), "y": float(parts[2]), "z": float(parts[3]), "size": 1.0})
    return verts


def из_клетки(cell_id: str) -> tuple[list[dict], list[dict], dict]:
    """Клетка → (исходные атомы, облако, meta) для round-trip."""
    cat = json.load(open(os.path.join(КЛЕТКИ, "каталог.json"), encoding="utf-8"))
    cells = next(v for v in cat.values() if isinstance(v, list))
    c = next(x for x in cells if x["id"] == cell_id)
    jp = os.path.join(КЛЕТКИ, c.get("атомы") or "")
    rec = json.load(open(jp, encoding="utf-8"))
    atoms = rec.get("atoms") or []
    облако, meta = атомы_в_облако(atoms, канон=False)
    return atoms, облако, meta


def совпадают(
    orig: Sequence[Mapping],
    recon: Sequence[Mapping],
    *,
    tol: float = 0.02,
) -> tuple[bool, float]:
    """Сравнение birth/freq/amp (допуск tol относительный)."""
    if len(orig) != len(recon):
        return False, math.inf

    def key(a):
        return (round(float(a.get("birth") or 0), 4), round(float(a.get("freq") or 0), 1))

    o = sorted(orig, key=key)
    r = sorted(recon, key=key)

    errs = []
    for a, b in zip(o, r):
        for field in ("birth", "freq", "amp"):
            va = float(a.get(field) or 0)
            vb = float(b.get(field) or 0)
            denom = max(abs(va), abs(vb), 1e-9)
            errs.append(abs(va - vb) / denom)
    worst = max(errs) if errs else 0.0
    return worst <= tol, worst

File: приложение/общее/образ_в_атомы.js
// Тринити V3 — обратная проекция: облако образа → скелеты атомов.
// Зеркало ядро/образ_в_атомы.py (единая логика с offline-тестами).

const X_SCALE = 1.8;
const Y_SCALE = 2.4;
const Z_SCALE = 1.2;
const T_ОБРАЗА = 3.0;
const F_LO = 60;
const F_HI = 9000;
const AMP_MIN = 0.05;
const AMP_MAX = 1.0;
const ОКНО_BIRTH = 0.05;
const MAX_СКЕЛЕТОВ = 24;
const K = 8;

function clip01(v) { return Math.max(0, Math.min(1, v)); }
function norm(v, a, b) { return b === a ? 0.5 : (v - a) / (b - a); }

export function границыАтомов(atoms) {
    const births = atoms.map(a => a.birth || 0);
    const lfs = atoms.map(a => Math.log10(Math.max(1, a.freq || 20)));
    const amps = atoms.map(a => a.amp || 0);
    return {
        tmin: Math.min(...births), tmax: Math.max(...births),
        fmin: Math.min(...lfs), fmax: Math.max(...lfs),
        amax: Math.max(...amps, 1e-6),
    };
}

export function атомыВОблако(atoms, { канон = false, tОбраза = T_ОБРАЗА } = {}) {
    if (!atoms?.length) return { облако: [], meta: {} };
    const b = границыАтомов(atoms);
    const meta = канон
        ? { режим: 'канон', t_образа: tОбраза }
        : { ...b, режим: 'физика', t_образа: tОбраза };
    const облако = atoms.map(a => {
        const birth = a.birth || 0;
        const freq = Math.max(1, a.freq || 20);
        const amp = a.amp || 0;
        let x, y, z;
        if (канон) {
            const nx = clip01(birth / tОбраза);
            x = (2 * nx - 1) * X_SCALE;
            const ny = clip01((Math.log(freq) - Math.log(F_LO)) / (Math.log(F_HI) - Math.log(F_LO)));
            y = (ny - 0.5) * Y_SCALE;
            z = clip01((amp - AMP_MIN) / (AMP_MAX - AMP_MIN)) * Z_SCALE;
        } else {
            x = (2 * norm(birth, b.tmin, b.tmax) - 1) * X_SCALE;
            y = (norm(Math.log10(freq), b.fmin, b.fmax) - 0.5) * Y_SCALE;
            z = (amp / b.amax) * Z_SCALE;
        }
        const pt = { x, y, z, size: a.size ?? Math.max(0.05, amp) };
        if (a.color_r != null) { pt.r = a.color_r; pt.g = a.color_g; pt.b = a.color_b; }
        return pt;
    });
    return { облако, meta };
}

function точкаВСкелет(pt, meta) {
    const { x, y, z } = pt;
    const w = pt.size ?? 1;
    let birth, freq, amp;
    if (meta.режим === 'физика' && meta.tmin != null) {
        birth = ((x / X_SCALE + 1) / 2) * (meta.tmax - meta.tmin) + meta.tmin;
        const logf = (y / Y_SCALE + 0.5) * (meta.fmax - meta.fmin) + meta.fmin;
        freq = 10 ** logf;
        amp = (z / Z_SCALE) * meta.amax;
    } else {
        const t = meta.t_образа ?? T_ОБРАЗА;
        birth = clip01((x / X_SCALE + 1) / 2) * t;
        const ny = clip01(y / Y_SCALE + 0.5);
        freq = Math.exp(ny * (Math.log(F_HI) - Math.log(F_LO)) + Math.log(F_LO));
        amp = AMP_MIN + clip01(z / Z_SCALE) * (AMP_MAX - AMP_MIN);
    }
    const sk = {
        birth: Math.max(0, birth),
        freq: Math.max(F_LO, Math.min(F_HI, freq)),
        amp: meta.режим === 'физика' ? Math.max(0, amp) : Math.max(AMP_MIN, Math.min(AMP_MAX, amp)),
        _w: w,
    };
    if (pt.r != null) sk.цвет = { r: pt.r, g: pt.g, b: pt.b };
    return sk;
}

function слитьОкно(group) {
    const wsum = group.reduce((s, g) => s + g._w, 0) || 1;
    return {
        birth: group.reduce((s, g) => s + g.birth * g._w, 0) / wsum,
        freq: group.reduce((s, g) => s + g.freq * g._w, 0) / wsum,
        amp: group.reduce((s, g) => s + g.amp * g._w, 0) / wsum,
        _w: wsum,
        ...(group[0].цвет ? { цвет: group[0].цвет } : {}),
    };
}

function ограничитьПлотность(скелеты) {
    const sorted = [...скелеты].sort((a, b) => a.birth - b.birth);
    const out = [];
    let i = 0;
    while (i < sorted.length) {
        const t0 = sorted[i].birth;
        const group = [];
        while (i < sorted.length && sorted[i].birth - t0 <= ОКНО_BIRTH + 1e-9) group.push(sorted[i++]);
        if (group.length <= MAX_СКЕЛЕТОВ) out.push(...group);
        else {
            group.sort((a, b) => b._w - a._w);
            out.push(...group.slice(0, MAX_СКЕЛЕТОВ));
            if (group.length > MAX_СКЕЛЕТОВ) out.push(слитьОкно(group.slice(MAX_СКЕЛЕТОВ)));
        }
    }
    return out;
}

function dist3(a, b) {
    return Math.hypot(a[0] - b[0], a[1] - b[1], a[2] - b[2]);
}

function локКривизна(pts, i) {
    if (pts.length < 3) return 0;
    const d = pts.map((p, j) => ({ j, d: j === i ? Infinity : dist3(p, pts[i]) }));
    d.sort((a, b) => a.d - b.d);
    const nn = d.slice(0, Math.min(K, pts.length - 1)).map(x => pts[x.j]);
    if (nn.length < 2) return 0;
    const cx = nn.reduce((s, p) => s + p[0], 0) / nn.length;
    const cy = nn.reduce((s, p) => s + p[1], 0) / nn.length;
    const cz = nn.reduce((s, p) => s + p[2], 0) / nn.length;
    let sxx = 0, syy = 0, szz = 0;
    for (const p of nn) {
        sxx += (p[0] - cx) ** 2;
        syy += (p[1] - cy) ** 2;
        szz += (p[2] - cz) ** 2;
    }
    const trace = (sxx + syy + szz) / nn.length;
    return trace > 1e-12 ? Math.min(sxx, syy, szz) / (sxx + syy + szz) : 0;
}

function локПлотность(pts, i) {
    if (pts.length <= 1) return 1;
    const d = pts.map((p, j) => j === i ? Infinity : dist3(p, pts[i])).sort((a, b) => a - b);
    const r = d.slice(0, Math.min(K, pts.length - 1)).reduce((s, v) => s + v, 0) / Math.min(K, pts.length - 1);
    return 1 / Math.max(r, 1e-3);
}

export function облакоВСкелеты(облако, { meta = null, tОбраза = T_ОБРАЗА, ограничитьПлотность = true } = {}) {
    if (!облако?.length) return [];
    const m = meta ? { ...meta } : { режим: 'канон', t_образа: tОбраза };
    if (!m.режим) m.режим = 'канон';
    let raw = облако.map(p => точкаВСкелет(p, m));
    if (ограничитьПлотность) raw = ограничитьПлотность(raw);
    const pts = raw.map(s => [s.birth, Math.log(Math.max(s.freq, 1)), s.amp]);
    return raw.map((s, i) => ({
        birth: s.birth,
        freq: s.freq,
        amp: s.amp,
        лок_кривизна: локКривизна(pts, i),
        лок_плотность: локПлотность(pts, i),
        ...(s.цвет ? { цвет: s.цвет } : {}),
    }));
}

export function образВАтомы(облако, opts) {
    return облакоВСкелеты(облако, opts);
}

/** Пиксели канваса → точки (y инвертирован, z = яркость). */
export function изКанваса(imageData, { порог = 30, шаг = 2 } = {}) {
    const { data, width, height } = imageData;
    const pts = [];
    for (let y = 0; y < height; y += шаг) {
        for (let x = 0; x < width; x += шаг) {
            const i = (y * width + x) * 4;
            const r = data[i], g = data[i + 1], b = data[i + 2], a = data[i + 3];
            const ярк = (r + g + b) / 3;
            if (a < 10 || ярк < порог) continue;
            pts.push({
                x: (x / width) * 2 * X_SCALE - X_SCALE,
                y: ((height - y) / height - 0.5) * Y_SCALE * 2,
                z: (ярк / 255) * Z_SCALE,
                size: ярк / 255,
                r, g, b,
            });
        }
    }
    return pts;
}

if (typeof window !== 'undefined') {
    window.__образ_в_атомы = { атомыВОблако, облакоВСкелеты, образВАтомы, изКанваса };
}
